README.md
[![Cookbook Version](https://img.shields.io/cookbook/v/spring-boot.svg)](https://supermarket.chef.io/cookbooks/spring-boot)
[![Build Status](https://travis-ci.org/goblin23/chef-spring-boot.svg?branch=master)](https://travis-ci.org/goblin23/chef-spring-boot)
=====================================================
Spring Boot Chef Cookbook
=====================================================
This Chef cookbook provides a [custom resource](https://docs.chef.io/custom_resources.html) **spring_boot_web_app** to install and deploy Spring Boot
Applications. Currently only supports Spring Boot Web
Apps.

Also, will only work on systems with `systemd` or `init.d` (System V) based init systems.

For `init.d`, the Spring Boot App must be configured
to be [fully executable](http://docs.spring.io/spring-boot/docs/current/reference/html/deployment-install.html).
Log location can also not be set for `init.d` and all
console logs will be written to `/var/log/<appname>.log`

This [custom resource](https://docs.chef.io/custom_resources.html) does not install java.
Java must be installed on the target node before using the [custom resource](https://docs.chef.io/custom_resources.html) **spring_boot_web_app**
To use it you have to include `depends 'spring-boot'` in your `metadata.rb`.

Syntax
=====================================================

This cookbook provides a [custom resource](https://docs.chef.io/custom_resources.html) named `spring_boot_web_app`.
The simplest use of the **spring_boot_web_app** resource is:
```
spring_boot_web_app 'name_of_webapp' do
  jar_remote_path 'http://example.com/path/to/your/jar/your_jar.jar'
end
```
which will download your artifact from **jar_remote_path** and create a "systemd" service listening on port ***8080***
started by the user ***bootapp***

The full syntax for all of the properties that are available to the **spring_boot_web_app** resource is:
```
spring_boot_web_app 'name_of_webapp' do
  notifies                   # see description
  user                       String # defaults to bootapp
  group                      String # defaults to bootapp
  port                       Integer # defaults to 8080
  jar_remote_path            String
  java_opts                  String
  boot_opts                  String
  properties                 Hash
  repo_user                  String
  repo_password              String
  init_system                String # defaults to systemd
  wait_for_http              [TrueClass, FalseClass] # defaults to true
  wait_for_http_retries      Integer # defaults 24
  wait_for_http_retry_delay  Integer # defaults 5
  subscribes                 # see description
  action                     Symbol # defaults to :install if not specified
end
```
where

*   ``user``  - the user that runs the webapp `default: 'bootapp'`
*   ``group`` - the group the user who runs the web_app belongs to `default: 'bootapp'``
*   ``port`` - the port that the web_app listens on `default: 8080`
*   ``jar_remote_path`` - the location the jar_file is fetched from
*   ``java_opts`` - the `JAVA_OPTS` the application is started with
*   ``boot_opts`` - the `BOOT_OPTS` the application is started with
*   ``properties`` - a Hash that describes properties files
*   ``repo_user`` - the user if your `jar_remote_path` is protected by basic auth
*   ``repo_password`` - the password if your `jar_remote_path` is protected by basic auth
*   ``init_system`` - for now `systemd` and `initd` are valid options `default: 'systemd'`
*   ``wait_for_http`` - should chef wait for the webapp to answer `default: true`
*   ``wait_for_http_retries`` - how many times should chef-client retry   `default: 24`
*   ``wait_for_http_retry_delay`` - how long should chef-client wait before each request `default: 5`
*   ``jmx_credentials`` - a hash that describes jmx_credentials

See "Properties" section below for more information about all of the properties that may be used with this resource.

Actions
=====================================================
This resource has the following actions:

``:install``
  Default. Installs spring_boot_web_app from `jar_remote_path`

``:uninstall``
  Removes previosly installed spring_boot_web_app instances

### Properties

``notifies``
   **Ruby Type:** Symbol, `'Chef::Resource[String]'`

   A resource may notify another resource to take action when its state changes. Specify a ``'resource[name]'``, the ``:action`` that resource should take, and then the ``:timer`` for that action. A resource may notify more than one resource; use a ``notifies`` statement for each resource to be notified.

   A timer specifies the point during the chef-client run at which a notification is run. The following timers are available:

   ``:before``
      Specifies that the action on a notified resource should be run before processing the resource block in which the notification is located.

   ``:delayed``
      Default. Specifies that a notification should be queued up, and then executed at the very end of the chef-client run.

   ``:immediate``, ``:immediately``
      Specifies that a notification should be run immediately, per resource notified.

   The syntax for ``notifies`` is:

```
notifies :action, 'resource[name]', :timer
```

``properties``
   **Ruby Type:** Hash

   Optional. The keys on the toplevel of the hash are the filename of a properties file postfixed with **.properties**
   the values of the toplevel keys are hashes containing key, value pairs that are written out to the file - for example:
```
{ 'app_1_initd' => { 'a' => '5', 'b' => '10' }, 'other_properties' => { 'c' => '25'} } 
```
would create two properties files:

file                          | content       |
------------                  | ------------- |
`app_1_initd.properties`      | `a=5` `b=10`  |
`other_properties.properties` | `c=25`        |

``jmx_credentials``
   **Ruby Type:** Hash

  Optional. the keys on the toplevel are the usernames their values are hases containing a `password` and `access` key.
  - for example:
```
{ 'monitorRole' => { 'password' => 'monitor', 'access' => 'readonly'} }
```
would create the two files jmxremote.access and jmxremote.password as follows:

```
# jmxremote.access
monitorRole readonly

# jmxremote.password
monitorRole monitor
```
Examples
=====================================================

Examples can be found in test/fixtures/cookbooks/fake/recipes/


CHANGELOG.md
spring-boot CHANGELOG
====
1.0.0
---
-   [JonsJava](https://github.com/jonsjava)
-   Cleaned up recipes, added new features that Chef now expects, broke example recipes into fake cookbook.
-   [goblin23](https://github.com/goblin23)
-   Converted LWRP into a custom ressource
-   Removed dependency to the java cookbook

1.1.0
---
-   [Khorchammar](https://github.com/Khorchammar)
-   Added jmx configuration

1.1.1
---
-   [JonsJava](https://github.com/jonsjava)
-   Very minor tweak to the README.md to ensure that the Multi-Dimensional array is Cookstyle-compliant. (spaces before closing `}`)

1.1.2
---
-   [frank-carnovale](https://github.com/frank-carnovale) Set current working directory in service definition template
-   [goblin23](https://github.com/goblin23) removed kind_of statements from custom ressource
1.1.4
---
-   [svenatwork] (https://github.com/svenatwork/) fix systemd service file template
1.1.5
---
-   [svenatwork] (https://github.com/svenatwork/) set logging locations for stdout and stderr


resources/web_app.rb
resource_name :spring_boot_web_app
default_action :install

property :user, String, default: 'bootapp'
property :shell, String, default: '/usr/sbin/nologin'
property :group, String, default: 'bootapp'
property :port, Integer, default: 8080
property :jar_remote_path, String, required: true
property :java_opts, String, default: ''
property :boot_opts, String, default: ''
property :properties, Hash
property :repo_user, String
property :repo_password, String
property :jmx_port, Integer
property :jmx_ssl, [true, false], default: false
property :jmx_credentials, Hash, default: {
  'monitorRole' => {
    'password' => '',
    'access' => 'readonly',
  },
}
property :init_system, String, default: 'systemd'
property :wait_for_http, [true, false], default: true
property :wait_for_http_retries, Integer, default: 24
property :wait_for_http_retry_delay, Integer, default: 5

action :install do
  jar_directory = "/opt/spring-boot/#{new_resource.name}"
  jar_path = jar_directory + '/' + new_resource.name + '.jar'
  logging_directory = jar_directory + '/logs'
  jmx_access_path = jar_directory + '/jmxremote.access'
  jmx_password_path = jar_directory + '/jmxremote.password'
  unless new_resource.repo_user.nil? || new_resource.repo_password.nil?
    basic_auth = "#{new_resource.repo_user}:#{new_resource.repo_password}"
  end
  declare_resource(:user, new_resource.user) do
    shell new_resource.shell
  end

  declare_resource(:group, new_resource.group) do
    append true
    members [new_resource.user]
  end

  directory jar_directory do
    owner new_resource.user
    group new_resource.group
    mode '0500'
    action :create
    recursive true
  end

  unless new_resource.jmx_port.nil?
    content_jmx_access = ''
    new_resource.jmx_credentials.each do |username, values|
      content_jmx_access << "#{username} #{values['access']}\n"
    end
    declare_resource(:file, jmx_access_path) do
      content content_jmx_access
      owner new_resource.user
      group new_resource.group
      mode '0400'
    end

    content_jmx_password = ''
    new_resource.jmx_credentials.each do |username, values|
      content_jmx_password << "#{username} #{values['password']}\n"
    end
    declare_resource(:file, jmx_password_path) do
      content content_jmx_password
      owner new_resource.user
      group new_resource.group
      mode '0400'
    end

    new_resource.java_opts << ' -Dcom.sun.management.jmxremote=true -Dcom.sun.management.jmxremote.port=' + new_resource.jmx_port.to_s
    new_resource.java_opts << if property_is_set?(:jmx_credentials)
                                " -Dcom.sun.management.jmxremote.password.file=#{jmx_password_path} -Dcom.sun.management.jmxremote.authenticate=true"
                              else
                                ' -Dcom.sun.management.jmxremote.authenticate=false'
                              end
    new_resource.java_opts << ' -Dcom.sun.management.jmxremote.ssl=' + new_resource.jmx_ssl.to_s
  end

  directory logging_directory do
    owner new_resource.user
    group new_resource.group
    mode '0700'
    action :create
    recursive true
  end

  remote_file jar_path do
    source new_resource.jar_remote_path
    owner new_resource.user
    group new_resource.group
    unless basic_auth.nil?
      headers('Authorization' => "Basic #{Base64.encode64(basic_auth).strip}")
    end
    mode '0500'
    action :create
    notifies :restart, "service[#{new_resource.name}]", :delayed
  end

  # TODO: Make JAR immutable and then make it removable in uninstall
  # execute 'Make jar immutable' do
  #   command "chattr +i #{jar_path}"
  #   user 'root'
  # end
  if property_is_set?(:properties)
    new_resource.properties.each do |key, value|
      file "#{jar_directory}/#{key}.properties" do
        content value.map { |k, v| "#{k}=#{v}" }.join("\n")
      end
    end
  end
  if new_resource.init_system == 'systemd'
    template "/etc/systemd/system/#{new_resource.name}.service" do
      source 'bootapp.service.erb'
      mode '0664'
      owner 'root'
      group 'root'
      cookbook 'spring-boot'
      variables(
        service_name: new_resource.name,
        user: new_resource.user,
        jar_path: jar_path,
        java_opts: new_resource.java_opts,
        boot_opts: new_resource.boot_opts,
        port: new_resource.port,
        logging_directory: logging_directory
      )
      notifies :restart, "service[#{new_resource.name}]", :delayed
      notifies :run, 'execute[systemctl_daemon_reload]', :immediately
    end

    execute 'systemctl_daemon_reload' do
      command 'systemctl daemon-reload'
      action :nothing
    end

  elsif new_resource.init_system == 'init.d'
    template "#{jar_directory}/#{new_resource.name}.conf" do
      source 'bootapp.conf.erb'
      mode '0400'
      owner 'root'
      group 'root'
      cookbook 'spring-boot'
      variables(
        jar_path: jar_path,
        java_opts: new_resource.java_opts,
        boot_opts: new_resource.boot_opts,
        port: new_resource.port,
        logging_directory: logging_directory
      )
      notifies :restart, "service[#{new_resource.name}]", :delayed
    end

    link "/etc/init.d/#{new_resource.name}" do
      to jar_path
    end
  else
    Chef::Application.fatal!("Invalid init system specified: #{new_resource.init_system}", 1)
  end

  service new_resource.name do
    action [:enable, :start]
  end

  execute "Ensure #{new_resource.name} web app is started up" do
    command "curl http://127.0.0.1:#{new_resource.port}"
    retries new_resource.wait_for_http_retries
    retry_delay new_resource.wait_for_http_retry_delay
    only_if { new_resource.wait_for_http }
  end
end

action :uninstall do
  service new_resource.name do
    action [:stop, :disable]
  end

  if new_resource.init_system == 'systemd'
    file "/etc/systemd/system/#{new_resource.name}.service" do
      action :delete
      notifies :run, 'execute[systemctl_daemon_reload]', :immediately
    end

    execute 'systemctl_daemon_reload' do
      command 'systemctl daemon-reload'
      action :nothing
    end
  elsif new_resource.init_system == 'init.d'
    link "/etc/init.d/#{new_resource.name}" do
      action :delete
    end
  else
    Chef::Application.fatal!("Invalid init system specified: #{init_system}", 1)
  end

  directory "/opt/spring-boot/#{new_resource.name}" do
    recursive true
    action :delete
    user 'root'
  end
end
